### ISSUE TITLE
Mac addres on multiple guest take the same value

### ISSUE THREAD

**Issue by tekmans** [josenk/vagrant-vmware-esxi#89]
Hi
I'm trying to build multiple machine 

For a single machine : create OK, name OK, mac adress defined OK

But for multiple machine : create OK, name OK, i had mac adress both defined with the second mac

IP addresses are fixed with mac address into a dhcp server as 

```
    host Box-1  { hardware ethernet 00:50:56:00:00:00; fixed-address 10.62.18.201; }
    host Box-2  { hardware ethernet 00:50:56:00:00:01; fixed-address 10.62.18.202; }

```

VagrantFile is like this, loop are used following your example
```
nodes = [
  { hostname: 'Box-1', mac:'00:50:56:00:00:00' },
  { hostname: 'Box-2', mac:'00:50:56:00:00:01' }
]

Vagrant.configure("2") do |config|
  nodes.each do |node|
    config.vm.define node[:hostname] do |config_machine|
      config_machine.vm.hostname = node[:hostname]
      config_machine.vm.box = "generic/debian10"
      config_machine.vm.box_check_update = true
      config_machine.vm.synced_folder('.', '/vagrant', type: 'rsync')
      ssh_pub_key = File.readlines("/root/.ssh/id_rsa.pub").first.strip
      config_machine.vm.provision 'shell', inline: 'mkdir -p /root/.ssh'
      config_machine.vm.provision 'shell', inline: "echo #{ssh_pub_key} >> /root/.ssh/authorized_keys"
      config_machine.vm.provision 'shell', inline: "echo #{ssh_pub_key} >> /home/vagrant/.ssh/authorized_keys", privileged: false
    end
    config.vm.provider :vmware_esxi do |esxi|
      esxi.debug = 'true ip vmx'
      esxi.esxi_hostname = 'X.X.X.X'
      esxi.esxi_username = 'user'
      esxi.esxi_password = 'password'
      esxi.esxi_disk_store = 'datastore1'
      esxi.esxi_virtual_network = ['Interne']
      esxi.guest_memsize = '2048'
      esxi.guest_numvcpus = '2'
      esxi.guest_mac_address = [node[:mac]]
    end
  end
end
```

Log 

```
==> AMBox-2: --- Avail DS vols   : ["datastore1"]
==> AMBox-1: --- Avail DS vols   : ["datastore1"]
==> AMBox-2: --- Avail Networks  : ["VM Network", "Internet", "Management Network", "Interne", "demopulse", "pulsev4"]
==> AMBox-2: ---   --- ESXi Summary ---
==> AMBox-2: --- ESXi host       : X.X.X.X
==> AMBox-2: --- Virtual Network : ["Interne"]
==> AMBox-2: --- Disk Store      : datastore1
==> AMBox-2: --- Resource Pool   : /
==> AMBox-2: ---  --- Guest Summary ---
==> AMBox-2: --- VM Name         : AMBox-2
==> AMBox-2: --- Box             : generic/debian10
==> AMBox-2: --- Box Ver         : 1.9.19
==> AMBox-2: --- Memsize (MB)    : 2048
==> AMBox-2: --- CPUS            : 2
==> AMBox-2: --- Mac Address     : ["00:50:56:00:00:01"]   <---------------------------------- is the same
==> AMBox-2: --- Guest OS type   : debian10-64
==> AMBox-2: ---   --- Guest Build ---
ovftool command: ovftool --noSSLVerify   --acceptAllEulas -dm=thin  -ds="datastore1" --name="AMBox-2" "/root/.vagrant.d/boxes/generic-VAGRANTSLASH-debian10/1.9.19/vmware_desktop/ZZZZ_AMBox-2.vmx" vi://user:******@X.X.X.X/
==> AMBox-1: --- Avail Networks  : ["VM Network", "Internet", "Management Network", "Interne", "demopulse", "pulsev4"]
==> AMBox-1: ---   --- ESXi Summary ---
==> AMBox-1: --- ESXi host       : X.X.X.X
==> AMBox-1: --- Virtual Network : ["Interne"]
==> AMBox-1: --- Disk Store      : datastore1
==> AMBox-1: --- Resource Pool   : /
==> AMBox-1: ---  --- Guest Summary ---
==> AMBox-1: --- VM Name         : AMBox-1
==> AMBox-1: --- Box             : generic/debian10
==> AMBox-1: --- Box Ver         : 1.9.19
==> AMBox-1: --- Memsize (MB)    : 2048
==> AMBox-1: --- CPUS            : 2
==> AMBox-1: --- Mac Address     : ["00:50:56:00:00:01"]   <---------------------------------- is the same
==> AMBox-1: --- Guest OS type   : debian10-64
==> AMBox-1: ---   --- Guest Build ---
```

First i had used reserved wmvare mac address (visible on vmware web client alert), and i had change to non reserved, but mac address are then same for both

Maybe I misunderstood something?

**Comment by josenk** [comment]
You did not create your Vagrantfile correctly...   Multimachine Vagranfiles can be pretty tricky.

Try this minimal example:
```
nodes = [
  { hostname: 'Box-1', mac:'00:50:56:00:00:00' },
  { hostname: 'Box-2', mac:'00:50:56:00:00:01' }
]

Vagrant.configure("2") do |config|
  nodes.each do |node|
    config.vm.define node[:hostname] do |node_config|
      node_config.vm.hostname = node[:hostname]
      node_config.vm.box = node[:box]
      node_config.vm.box = "generic/debian10"

      config.vm.provider :vmware_esxi do |esxi|
        esxi.esxi_hostname = 'x.x.x.x'
        esxi.esxi_password = 'File:'
        esxi.guest_mac_address = [node[:mac]]
      end
    end
  end
end
```

**Comment by tekmans** [comment]
Hi thanks for the reply but is still the same


```
# -*- mode: ruby -*-
# vi: set ft=ruby :

nodes = [
  { hostname: 'M-pulse', box:'generic/debian10', mac:'00:50:56:00:10:00', ostype:'centos7-64' },
  { hostname: 'M-relai-1', box:'generic/centos7', mac:'00:50:56:00:11:00', ostype:'debian10-64' }
]


Vagrant.configure("2") do |config|
  nodes.each do |node|
    config.vm.define node[:hostname] do |node_config|
      node_config.vm.hostname = node[:hostname]
      node_config.vm.box = node[:box]

      config.vm.provider :vmware_esxi do |esxi|
        esxi.debug = 'true ip vmx'
        esxi.esxi_hostname = 'X.X.X.X'
        esxi.esxi_username = 'user'
        esxi.esxi_password = 'password'
        esxi.esxi_disk_store = 'datastore1'
        esxi.guest_mac_address = [node[:mac]] <------- it needs bracket around to avoid invalid
        esxi.guest_guestos = node[:ostype] <---------- i had tried also with type of data 
      end
    end
  end
end
```
```
# vagrant up

RUBY_PLATFORM: x86_64-linux
Testing esxi connectivity
==> M-pulse: --- ESXi version    : VMware ESXi 6.5.0 build-5969303
RUBY_PLATFORM: x86_64-linux
Testing esxi connectivity
Bringing machine 'M-pulse' up with 'vmware_esxi' provider...
Bringing machine 'M-relai-1' up with 'vmware_esxi' provider...
==> M-relai-1: Virtual Machine will be built.
==> M-pulse: Virtual Machine will be built.
VMware ovftool 4.3.0 (build-12320924)
VMware ovftool 4.3.0 (build-12320924)
==> M-pulse: --- Avail DS vols   : ["datastore1"]
==> M-relai-1: --- Avail DS vols   : ["datastore1"]
==> M-pulse: --- Avail Networks  : ["VM Network", "Internet", "Management Network", "Interne"]
==> M-pulse: --- WARNING         : esxi_virtual_network[0] not set, using VM Network
==> M-pulse: ---   --- ESXi Summary ---
==> M-pulse: --- ESXi host       : X.X.X.X
==> M-pulse: --- Virtual Network : ["VM Network"]
==> M-pulse: --- Disk Store      : datastore1
==> M-pulse: --- Resource Pool   : /
==> M-pulse: ---  --- Guest Summary ---
==> M-pulse: --- VM Name         : M-pulse
==> M-pulse: --- Box             : generic/debian10
==> M-pulse: --- Box Ver         : 1.9.19
==> M-pulse: --- Memsize (MB)    : 2048
==> M-pulse: --- CPUS            : 2
==> M-pulse: --- Mac Address     : ["00:50:56:00:11:00"] <--- is always the second
==> M-pulse: --- Guest OS type   : debian10-64 <------------- is always the second
==> M-pulse: ---   --- Guest Build ---
ovftool command: ovftool --noSSLVerify   --acceptAllEulas -dm=thin  -ds="datastore1" --name="M-pulse" "/root/.vagrant.d/boxes/generic-VAGRANTSLASH-debian10/1.9.19/vmware_desktop/ZZZZ_M-pulse.vmx" vi://user:******@X.X.X.X/
==> M-relai-1: --- Avail Networks  : ["VM Network", "Internet", "Management Network", "Interne"]
==> M-relai-1: --- WARNING         : esxi_virtual_network[0] not set, using VM Network
==> M-relai-1: ---   --- ESXi Summary ---
==> M-relai-1: --- ESXi host       : X.X.X.X
==> M-relai-1: --- Virtual Network : ["VM Network"]
==> M-relai-1: --- Disk Store      : datastore1
==> M-relai-1: --- Resource Pool   : /
==> M-relai-1: ---  --- Guest Summary ---
==> M-relai-1: --- VM Name         : M-relai-1
==> M-relai-1: --- Box             : generic/centos7
==> M-relai-1: --- Box Ver         : 1.9.18
==> M-relai-1: --- Memsize (MB)    : 2048
==> M-relai-1: --- CPUS            : 2
==> M-relai-1: --- Mac Address     : ["00:50:56:00:11:00"] <--- is always the second
==> M-relai-1: --- Guest OS type   : debian10-64 <------------- is always the second
==> M-relai-1: ---   --- Guest Build ---

```

It looks as the esxi part is not built to be multi guest because it always take the lastest information.
If i had tried with tree guest, all information in esxi part are the lastest.

**Comment by josenk** [comment]
My example works...   Here is an expanded version that includes number of cpus...
```
nodes = [
  { hostname: 'Box-1', box:'generic/debian10', mac:'00:50:56:00:00:00', numvcpus:'1' },
  { hostname: 'Box-2', box:'generic/centos7', mac:'00:50:56:00:00:01', numvcpus:'2' }
]

Vagrant.configure("2") do |config|
  nodes.each do |node|
    config.vm.define node[:hostname] do |node_config|
      node_config.vm.hostname = node[:hostname]
      node_config.vm.box = node[:box]

      config.vm.provider :vmware_esxi do |esxi|
        esxi.esxi_hostname = 'esxi'
        esxi.esxi_password = 'File:'
        esxi.guest_mac_address = [node[:mac]]
        esxi.guest_numvcpus = node[:numvcpus]
      end
    end
  end
end


# vagrant up

Bringing machine 'Box-1' up with 'vmware_esxi' provider...
Bringing machine 'Box-2' up with 'vmware_esxi' provider...
==> Box-1: Virtual Machine will be built.
==> Box-2: Virtual Machine will be built.
VMware ovftool 4.3.0 (build-7948156)
VMware ovftool 4.3.0 (build-7948156)
==> Box-2: --- WARNING         : esxi_disk_store not set, using "--- Least Used ---"
==> Box-1: --- WARNING         : esxi_disk_store not set, using "--- Least Used ---"
==> Box-1: --- WARNING         : esxi_virtual_network[0] not set, using 192.168.1
==> Box-1: ---   --- ESXi Summary ---
==> Box-1: --- ESXi host       : esxi
==> Box-1: --- Virtual Network : ["192.168.1"]
==> Box-1: --- Disk Store      : DS_4TB
==> Box-1: --- Resource Pool   : /
==> Box-1: ---  --- Guest Summary ---
==> Box-1: --- VM Name         : Box-1
==> Box-1: --- Box             : generic/debian10
==> Box-1: --- Box Ver         : 1.9.20
==> Box-1: --- Memsize (MB)    : 2048
==> Box-1: --- CPUS            : 1                          <------  This is correct
==> Box-1: --- Mac Address     : ["00:50:56:00:00:00"]    <------  This is correct
==> Box-1: --- Guest OS type   : debian8-64
==> Box-1: ---   --- Guest Build ---
==> Box-2: --- WARNING         : esxi_virtual_network[0] not set, using 192.168.1
==> Box-2: ---   --- ESXi Summary ---
==> Box-2: --- ESXi host       : 192.168.1.195
==> Box-2: --- Virtual Network : ["192.168.1"]
==> Box-2: --- Disk Store      : DS_4TB
==> Box-2: --- Resource Pool   : /
==> Box-2: ---  --- Guest Summary ---
==> Box-2: --- VM Name         : Box-2  
==> Box-2: --- Box             : generic/centos7   
==> Box-2: --- Box Ver         : 1.9.14
==> Box-2: --- Memsize (MB)    : 2048
==> Box-2: --- CPUS            : 2                              <------  This is correct
==> Box-2: --- Mac Address     : ["00:50:56:00:00:01"]     <------  This is correct
==> Box-2: --- Guest OS type   : centos-64
==> Box-2: ---   --- Guest Build ---
```

**Comment by tekmans** [comment]
I have tried your file except ip and password the results is the same

I have also downgraded the ovf tools to have the same build as yours and the result still the same

```
vmware-installer -l
Product Name         Product Version     
==================== ====================
vmware-ovftool       4.3.0.7948156  
```
```
# cat Vagrantfile
# -*- mode: ruby -*-
# vi: set ft=ruby :

nodes = [
  { hostname: 'Box-1', box:'generic/debian10', mac:'00:50:56:00:00:00', numvcpus:'1' },
  { hostname: 'Box-2', box:'generic/centos7', mac:'00:50:56:00:00:01', numvcpus:'2' }
]

Vagrant.configure("2") do |config|
  nodes.each do |node|
    config.vm.define node[:hostname] do |node_config|
      node_config.vm.hostname = node[:hostname]
      node_config.vm.box = node[:box]

      config.vm.provider :vmware_esxi do |esxi|
        esxi.esxi_hostname = 'X.X.X.X'
        esxi.esxi_username = 'username'
        esxi.esxi_password = 'password'
        esxi.guest_mac_address = [node[:mac]]
        esxi.guest_numvcpus = node[:numvcpus]
      end
    end
  end
end

```
```
# vagrant up

Bringing machine 'Box-1' up with 'vmware_esxi' provider...
Bringing machine 'Box-2' up with 'vmware_esxi' provider...
==> Box-2: Virtual Machine will be built.
==> Box-1: Virtual Machine will be built.
VMware ovftool 4.3.0 (build-7948156)
VMware ovftool 4.3.0 (build-7948156)
==> Box-1: --- WARNING         : esxi_disk_store not set, using "--- Least Used ---"
==> Box-2: --- WARNING         : esxi_disk_store not set, using "--- Least Used ---"
==> Box-1: --- WARNING         : esxi_virtual_network[0] not set, using VM Network
==> Box-1: ---   --- ESXi Summary ---
==> Box-1: --- ESXi host       : X.X.X.X
==> Box-1: --- Virtual Network : ["VM Network"]
==> Box-1: --- Disk Store      : datastore1
==> Box-1: --- Resource Pool   : /
==> Box-1: ---  --- Guest Summary ---
==> Box-1: --- VM Name         : Box-1
==> Box-1: --- Box             : generic/debian10
==> Box-1: --- Box Ver         : 1.9.19
==> Box-1: --- Memsize (MB)    : 2048
==> Box-1: --- CPUS            : 2
==> Box-1: --- Mac Address     : ["00:50:56:00:00:01"]
==> Box-1: --- Guest OS type   : debian10-64
==> Box-1: ---   --- Guest Build ---
==> Box-2: --- WARNING         : esxi_virtual_network[0] not set, using VM Network
==> Box-2: ---   --- ESXi Summary ---
==> Box-2: --- ESXi host       : X.X.X.X
==> Box-2: --- Virtual Network : ["VM Network"]
==> Box-2: --- Disk Store      : datastore1
==> Box-2: --- Resource Pool   : /
==> Box-2: ---  --- Guest Summary ---
==> Box-2: --- VM Name         : Box-2
==> Box-2: --- Box             : generic/centos7
==> Box-2: --- Box Ver         : 1.9.18
==> Box-2: --- Memsize (MB)    : 2048
==> Box-2: --- CPUS            : 2
==> Box-2: --- Mac Address     : ["00:50:56:00:00:01"]
==> Box-2: --- Guest OS type   : debian10-64
==> Box-2: ---   --- Guest Build ---

```

Could it be something about version ?
```
# vagrant -v
Vagrant 2.2.5

# vagrant plugin list
vagrant-reload (0.0.1, global)
vagrant-vmware-esxi (2.4.4, global)
vagrant-winrm-syncedfolders (1.0.1, global)

```

**Comment by josenk** [comment]
When I upgraded from Vagrant 2.2.1 to 2.2.5 I was able to reproduce your problem.   The cpu and MAC address was duplicated in both VMs.

I updated my development environment to 2.2.6.dev, but it seemed to have messed things up!   I'm unable to build.     I get the exact same error as here. https://github.com/vagrant-libvirt/vagrant-libvirt/issues/1018

I'll try to get this straightened out.   In the mean time, you may want to try a previous version of vagrant to see if it resolves your problems.

**Comment by josenk** [comment]
Using your example Vagrantfile, I have the same problem with the 2.2.6.dev (development version).  If I downgrade to 2.2.3, it works fine.

I found this...  
>https://github.com/hashicorp/vagrant/issues/10957

I think the format might have change for multimachines.   Seems others have the same problem with 2.2.4 & 2.2.5.   They give an example minimal Vagrantfile.

**Comment by MediumBrownToast** [comment]
I'm a Vagrant newbie so my apologies if this sounds odd.

I'm using 2.2.6 and experiencing the same issue. Is it because the variables are lazy-loaded as stated here? https://www.vagrantup.com/docs/vagrantfile/tips.html#loop-over-vm-definitions

**Comment by josenk** [comment]
I updated the multimachine example to use the correct format.   I believe you are right MediumBrownToast that there was a lazy-loaded problem with the loop.  (Interesting is older versions of Vagrant tolerated it OK...)  The new example doesn't have the MAC address specifically, but I did test it on my system successfully with Vagrant 2.2.7

> https://github.com/josenk/vagrant-vmware-esxi/blob/master/example_box/Vagrantfile-multimachine

**Comment by MediumBrownToast** [comment]
The new example doesn't have the MAC address specifically, but I did test it on my system successfully with Vagrant 2.2.7
> 
> > https://github.com/josenk/vagrant-vmware-esxi/blob/master/example_box/Vagrantfile-multimachine

Thanks for that.

It's been a while since I used Vagrant so I can't quite remember and I don't have access to my ESXi machine atm, but could MAC addresses now be successfully specified for each VM?

It'd be useful as my DHCP server could assign IPs based on MAC.
